src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -0,0 +1,644 @@
+        case m: vpr.Function => Vertex.Function(m.name)
+        case m: vpr.Predicate => Vertex.PredicateBody(m.name)
+        case m: vpr.Field => Vertex.Field(m.name)
+        case m: vpr.Domain => Vertex.DomainType(vpr.DomainType(domain = m, (m.typVars zip m.typVars).toMap))
+        case _: vpr.ExtensionMember => ???
+      }
+    }
+
+    /** Returns the subprogram induced by the set of vertices. */
+    def inverse(program: vpr.Program): Set[Vertex] => vpr.Program = {
+      val methodTable = program.methods.map(n => (n.name, n)).toMap
+      val functionTable = program.functions.map(n => (n.name, n)).toMap
+      val predicateTable = program.predicates.map(n => (n.name, n)).toMap
+      val fieldTable = program.fields.map(n => (n.name, n)).toMap
+      val domainTable = program.domains.map(n => (n.name, n)).toMap
+      val domainFunctionTable = program.domains.flatMap(d => d.functions.map(f => (f.name, (f, d)))).toMap
+
+      { (vertices: Set[Vertex]) =>
+
+        val methods = {
+          val ms = vertices.collect{ case v: Method => methodTable(v.methodName) }
+          val stubs = vertices.collect{ case v: MethodSpec => val m = methodTable(v.methodName); m.copy(body = None)(m.pos, m.info, m.errT) }
+          val filteredStubs = stubs.filterNot(stub => ms.exists(_.name == stub.name))
+          (ms ++ filteredStubs).toSeq
+        }
+        val funcs = vertices.collect{ case v: Function => functionTable(v.functionName) }.toSeq
+        val preds = {
+          val psigs = vertices.collect{ case v: PredicateSig => val p = predicateTable(v.predicateName); p.copy(body = None)(p.pos, p.info, p.errT) }.toSeq
+          val pbodies = vertices.collect{ case v: PredicateBody => predicateTable(v.predicateName) }.toSeq
+          val filteredSigs = psigs.filterNot(sig => pbodies.exists(_.name == sig.name))
+          pbodies ++ filteredSigs
+        }
+        val fields = vertices.collect{ case v: Field => fieldTable(v.fieldName) }.toSeq
+        val domains = {
+          val fs = vertices.collect{ case v: DomainFunction => domainFunctionTable(v.funcName) }.toSeq.groupMap(_._2)(_._1)
+          val as = vertices.collect{ case v: DomainAxiom => (v.v, v.d)}.toSeq.groupMap(_._2)(_._1)
+          val ds = vertices.collect{ case v: DomainType => domainTable(v.v.domainName) }.toSeq
+          val totalDs = (ds ++ fs.keys ++ as.keys).distinct
+
+          totalDs.map{ d =>
+            d.copy(functions = fs.getOrElse(d, Seq.empty), axioms = as.getOrElse(d, Seq.empty))(d.pos, d.info, d.errT)
+          }
+        }
+
+        program.copy(
+          methods = methods,
+          functions = funcs,
+          predicates = preds,
+          fields = fields,
+          domains = domains
+        )(program.pos, program.info, program.errT)
+      }
+    }
+  }
+
+  type Edge[T] = (T, T)
+  object Edges {
+
+    /** Returns all dependencies induced by a member. */
+    def dependencies(member: vpr.Member): Seq[Edge[Vertex]] = {
```

Comment:
Why do you use `Seq` instead of `Set`? Does ordering have some semantics or offer better performance?

Reply:
I am ordering at a later point in time, where it is faster.

Reply:
I see, so you're using a sequence here such that you do not need to convert e.g. a set to a sequence later on

Reply:
Almost. It is about the vertices. Checking that vertices are distinct is more expensive than checking that integers are distinct. And since I am converting to integers either way, I removing the duplicates also at that point.